pkg/runner/cluster_swarm.go

```diff
@@ -1 +1,318 @@
 package runner
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"reflect"
+	"time"
+
+	"github.com/ipfs/testground/pkg/api"
+	"github.com/ipfs/testground/pkg/aws"
+	"github.com/ipfs/testground/pkg/logging"
+	"github.com/ipfs/testground/pkg/util"
+	"github.com/ipfs/testground/sdk/runtime"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/api/types/swarm"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stdcopy"
+
+	"github.com/imdario/mergo"
+)
+
+var (
+	_ api.Runner = &ClusterSwarmRunner{}
+)
+
+// ClusterSwarmRunnerConfig is the configuration object of this runner. Boolean
+// values are expressed in a way that zero value (false) is the default setting.
+type ClusterSwarmRunnerConfig struct {
+	// LogLevel sets the log level in the test containers (default: not set).
+	LogLevel string `toml:"log_level"`
+
+	// Background avoids tailing the output of containers, and displaying it as
+	// log messages (default: true).
+	Background bool `toml:"background"`
+
+	// DockerEndpoint is the URL of the docker swarm manager endpoint, e.g.
+	// "tcp://manager:2376"
+	DockerEndpoint string `toml:"docker_endpoint"`
+
+	// DockerTLS indicates whether client TLS is enabled.
+	DockerTLS bool `toml:"docker_tls"`
+
+	// DockerTLSCACertPath is the path to the CA Certificate. Only used if
+	// DockerTLS = true.
+	DockerTLSCACertPath string `toml:"docker_tls_ca_cert_path"`
+
+	// DockerTLSCertPath is the path to our client cert, signed by the CA. Only
+	// used if DockerTLS = true.
+	DockerTLSCertPath string `toml:"docker_tls_cert_path"`
+
+	// DockerTLSKeyPath is our private key. Only used if DockerTLS = true.
+	DockerTLSKeyPath string `toml:"docker_tls_key_path"`
+
+	// RemoveService removes the service after all instances have finished and
+	// all logs have been piped. Only used when running in foreground mode
+	// (background = false).
+	RemoveService bool `toml:"rm_service"`
+}
+
+// defaultClusterSwarmConfig is the default configuration. Incoming configurations will be
+// merged with this object.
+var defaultClusterSwarmConfig = ClusterSwarmRunnerConfig{
+	Background:    false,
+	RemoveService: true,
+}
+
+// ClusterSwarmRunner is a runner that creates a Docker service to launch as
+// many replicated instances of a container as the run job indicates.
+type ClusterSwarmRunner struct{}
+
+// TODO runner option to keep containers alive instead of deleting them after
+// the test has run.
+func (*ClusterSwarmRunner) Run(input *api.RunInput) (*api.RunOutput, error) {
```

Comment:
No context?

Reply:
Not for now. But we should use contexts: https://github.com/ipfs/testground/issues/131.